File: app/src/main/java/com/unsupportedpastels/hermesandroid/relay/RelayDeviceIdentity.kt
package com.unsupportedpastels.hermesandroid.relay

import android.content.Context
import android.os.Build
import android.provider.Settings

/**
 * The name this phone reports to a relay host in its admission envelope, so
 * the host's device list can show "Mark's Fold" instead of a fingerprint.
 * The user's own device name wins; the model is the fallback.
 */
object RelayDeviceIdentity {
    @Volatile
    var name: String? = null
        private set

    fun initialize(context: Context) {
        val userName = runCatching {
            Settings.Global.getString(context.contentResolver, Settings.Global.DEVICE_NAME)
        }.getOrNull()?.trim().orEmpty()
        val model = listOf(Build.MANUFACTURER, Build.MODEL)
            .filter { !it.isNullOrBlank() }
            .joinToString(" ")
            .trim()
        name = userName.ifEmpty { model }.ifEmpty { null }
    }
}

File: ios/Mercury/MercuryKit/Relay/RelayDeviceIdentity.swift
import Foundation
import UIKit

/// The name this phone reports to a relay host in its admission envelope, so
/// the host's device list can show the phone by name instead of a fingerprint.
/// iOS returns a generic "iPhone" for the user-set name on recent systems, so
/// the model name is appended when the two differ.
enum RelayDeviceIdentity {
    static let name: String? = {
        let device = UIDevice.current
        let userName = device.name.trimmingCharacters(in: .whitespacesAndNewlines)
        let model = device.model.trimmingCharacters(in: .whitespacesAndNewlines)
        let combined: String
        if userName.isEmpty { combined = model }
        else if userName.caseInsensitiveCompare(model) == .orderedSame { combined = model }
        else { combined = userName }
        return combined.isEmpty ? nil : combined
    }()
}
